pkg/ccl/sqlproxyccl/proxy_handler.go

```diff
@@ -441,6 +460,53 @@ func (handler *proxyHandler) handle(ctx context.Context, incomingConn *proxyConn
 	}
 }
 
+// handleCancelRequest handles a pgwire query cancel request by either
+// forwarding it to a SQL node or to another proxy.
+func (handler *proxyHandler) handleCancelRequest(cr *proxyCancelRequest) error {
+	if ci, ok := handler.cancelInfoMap.getCancelInfo(cr.SecretKey); ok {
+		if !ci.clientAddr.(*net.TCPAddr).IP.Equal(cr.ClientIP) {
+			// If the IP associated with the cancelInfo does not match the IP from
+			// which the request came, then ignore it.
+			return errors.Errorf("mismatched client IP for cancel request")
+		}
+		cancelConn, err := net.DialTimeout("tcp", ci.crdbAddr.String(), 5*time.Second)
+		if err != nil {
+			return err
+		}
+		defer cancelConn.Close()
+		if err := cancelConn.SetDeadline(timeutil.Now().Add(5 * time.Second)); err != nil {
+			return err
+		}
+		crdbRequest := &pgproto3.CancelRequest{
+			ProcessID: ci.origBackendKeyData.ProcessID,
+			SecretKey: ci.origBackendKeyData.SecretKey,
+		}
+		buf := crdbRequest.Encode(nil /* buf */)
+		if _, err := cancelConn.Write(buf); err != nil {
+			return err
+		}
+		if _, err := cancelConn.Read(buf); err != io.EOF {
+			return err
+		}
+	}
+	forwardConn, err := net.DialTimeout("tcp", cr.ProxyIP.String()+":26257", 5*time.Second)
```

Comment:
I think we can probably live with this, but just wanted to point out that with my recent changes to do graceful shutdowns, a sqlproxy container can live for up to an hour with it's tcp listener on port `26257` closed but still have open connections and be making new queries. If any of those longer lived connections try to cancel something, this dialer will fail to connect.

The SQL Proxy continues listening on port 8080 for HTTP requests, so theoretically we could have an HTTP handler for these cancellation requests and those would continue to work even after the proxy stops its TCP listener.

Reply:
Here's where the HTTP request are handled, the `mux` is created near the top of the file:

https://github.com/cockroachdb/cockroach/blob/7af6c0d3e2085cafda90d5f0b15d06f6a360f154/pkg/ccl/sqlproxyccl/server.go#L121